devdoc/channel.md

```diff
@@ -0,0 +1,267 @@
+
+**SRS_CHANNEL_43_010: [** `channel_pull` shall call `interlocked_add` to obtain the current `state` of the `channel`. **]**
+
+**SRS_CHANNEL_43_011: [** If the current `state` is `CHANNEL_PULLING` or `CHANNEL_PULLED`, `channel_pull` shall fail and return `CHANNEL_RESULT_REFUSED`. **]**
+
+**SRS_CHANNEL_43_059: [** If the current `state` is `CHANNEL_PUSHING`, `channel_pull` shall wait for the `state` to change by calling `wait_on_address`. **]**
+
+**SRS_CHANNEL_43_012: [** If the current `state` is `CHANNEL_PUSHED`, `channel_pull` shall : **]**
+
+- **SRS_CHANNEL_43_060: [** store `push_callback`, `push_context` and `data` given to `channel_push`. **]**
+
+- **SRS_CHANNEL_43_042: [** call `interlocked_compare_exchange` to increment the `EPOCH` and change the state to `CHANNEL_READY`. **]**
+
+- **SRS_CHANNEL_43_013: [** call the given `pull_callback` with `pull_context` as the given `pull_context`, `data` as the stored `data` and `result` as `CHANNEL_RESULT_OK`. **]**
+
+- **SRS_CHANNEL_43_053: [** release the reference to the stored `data` by calling `THANDLE_ASSIGN`. **]**
+
+- **SRS_CHANNEL_43_014: [** call the stored `push_callback` with the stored `push_context` and `result` as `CHANNEL_RESULT_OK`. **]**
+
+- **SRS_CHANNEL_43_054: [** set `out_op_pull` to `NULL` by calling `THANDLE_ASSIGN`. **]**
+
+- **SRS_CHANNEL_43_015: [** return `CHANNEL_RESULT_SYNC`. **]**
+
+**SRS_CHANNEL_43_016: [** If the current state is `CHANNEL_READY`, `channel_pull` shall: **]**
+
+**SRS_CHANNEL_43_061: [** set the state to `CHANNEL_PULLING` by calling `interlocked_compare_exchange`. **]**
+
+- **SRS_CHANNEL_43_017: [** call `async_op_create` with `cancel_pull` as the cancel function. **]**
+
+- **SRS_CHANNEL_43_018: [** store the given `pull_callback` and the given `pull_context` in the `channel`. **]**
+
+- **SRS_CHANNEL_43_019: [** store `state` in the created `THANDLE(ASYNC_OP)`. **]**
+
+- **SRS_CHANNEL_43_020: [** store the given `channel` in the created `THANDLE(ASYNC_OP)` by calling `THANDLE_INITIALIZE`. **]**
+
+- **SRS_CHANNEL_43_021: [** store the created `THANDLE(ASYNC_OP)` in `out_op_pull` by calling `THANDLE_INITIALIZE_MOVE`. **]**
+
+- **SRS_CHANNEL_43_062: [** set the `state` of the given `channel` to `CHANNEL_PULLED` by calling `interlocked_exchange`. **]**
+
+- **SRS_CHANNEL_43_063: [** signal waiting threads by calling `wake_by_address_all`. **]**
+
+- **SRS_CHANNEL_43_022: [** return `CHANNEL_RESULT_ASYNC`. **]**
+
+**SRS_CHANNEL_43_023: [** If there are any failures, `channel_pull` shall fail and return `CHANNEL_RESULT_ERROR`. **]**
+
+### cancel_pull
+```
+static void cancel_pull(void* context)
+```
+
+`cancel_pull` is the cancel function given to `async_op_create` when `channel_pull` is called.
+
+**SRS_CHANNEL_43_046: [** If the current `EPOCH` and `state` of the `channel` is the same as the initial `EPOCH` and `state` of the `channel` when `channel_pull` was called, `cancel_pull` shall: **]**
```

Comment:
>`state` of the `channel` [](http://example.com/codeflow?start=53&length=24)

why does the state of the channel matter for this cancel?

I can almost understand `epoch`, but ... state?

Reply:
Only epoch really needs to be compared. 
If epoch matches, state must match because once channel is in PULLED/PUSHED states, the only way to change the state is push/pull or cancel, both of which increment epoch.

Epoch must be compared because channel could cycle back to the PULLED state for a later operation, which means the current operation has already been completed or cancelled and nothing more needs to be done in the current call to cancel.